diff --git a/main.go b/main.go
index 56ab5dc..d011399 100644
--- a/main.go
+++ b/main.go
@@ -16,6 +16,8 @@ package main
 
 import (
 	"os"
+	"os/exec"
+	"time"
 
 	"github.com/mendersoftware/mender/app"
 	"github.com/mendersoftware/mender/cli"
@@ -23,6 +25,21 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
+// All this code is simply stolen from the
+// client reboot routine.
+func rebootClient() {
+	err := exec.Command("reboot").Run()
+	if err != nil {
+		os.Exit(1)
+	}
+
+	// Wait up to ten minutes for reboot to kill the client, otherwise the
+	// client may mistake a successful return code as "reboot is complete,
+	// continue". *Any* return from this function is an error.
+	time.Sleep(10 * time.Minute)
+	fmt.Println("System did not reboot, even though 'reboot' call succeeded.")
+}
+
 func doMain() int {
 	if err := cli.SetupCLI(os.Args); err != nil {
 		switch err {
@@ -43,5 +60,25 @@ func doMain() int {
 }
 
 func main() {
-	os.Exit(doMain())
+	// The client panics instead of rebooting, so that we can capture the
+	// coverage logs, before rebooting.
+	defer func() {
+		if r := recover(); r != nil {
+			if s, ok := r.(string); ok && s == "Client needs reboot!" {
+				coverReport()
+				rebootClient()
+			} else {
+				panic(r)
+			}
+		}
+	}()
+	// Set the path in which to store the coverage analysis files
+	os.Setenv("COVERAGE_FILEPATH", "/data/mender/")
+
+	ret := doMain()
+
+	coverReport() // Manually create the coverage report before exiting
+
+	os.Exit(ret)
+
 }
diff --git a/system/system.go b/system/system.go
index a96d7d8..b4ceb1a 100644
--- a/system/system.go
+++ b/system/system.go
@@ -17,9 +17,7 @@ package system
 import (
 	"os"
 	"os/exec"
-	"time"
 
-	"github.com/pkg/errors"
 )
 
 type SystemRebootCmd struct {
@@ -33,16 +31,7 @@ func NewSystemRebootCmd(command Commander) *SystemRebootCmd {
 }
 
 func (s *SystemRebootCmd) Reboot() error {
-	err := s.command.Command("reboot").Run()
-	if err != nil {
-		return err
-	}
-
-	// Wait up to ten minutes for reboot to kill the client, otherwise the
-	// client may mistake a successful return code as "reboot is complete,
-	// continue". *Any* return from this function is an error.
-	time.Sleep(10 * time.Minute)
-	return errors.New("System did not reboot, even though 'reboot' call succeeded.")
+	panic("Client needs reboot!")
 }
 
 type Commander interface {
